Extensible RetainableByteBufferPool  #6322

gregw opened the pull request:
Improves on #6332 for #6322

Add a Log2 bucket size imple for ArrayRetainableByteBufferPool

Even if we don't actually add/use the `LogBucket` class, I think the changes made to make the ArrayRetainableByteBufferPool extensible are good ones to have.

Signed-off-by: Greg Wilkins <gregw@webtide.com>

lorban committed:
use RetainableByteBuffer

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
#6322 Use RetainableByteBuffer and write a new pool for it

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
fix checkstyle

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
replace toDetailString with straight RetainableByteBuffer.toString

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
use RetainableByteBuffer for encrypted input buffer

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
add managed attributes

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
add a bit more monitoring to ArrayByteBufferPool

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
modify SslConnectionFactory to add ability to use Pool<RetainableByteBuffer> for encrypted input

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
remove unneeded counters

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
first shot at RetainableByteBuffer pool

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
cleanup acquire/retain mechanism

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
use pool-specific ctor

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
add optional ability to use RetainableByteBufferPool on the server

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
revert old ByteBufferPool changes

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
fix compilation error

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
register RetainableByteBufferPool bean in AbstractConnector

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
add RetainableByteBufferPool registration use in websocket

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
add RetainableByteBufferPool registration use in client

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
add RetainableByteBufferPool registration use in upgrade

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
make RetainableByteBufferPool never null

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
make RetainableByteBufferPool never null

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
make H2 connection use RetainableByteBuffer pool

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
make H2 connection use RetainableByteBuffer pool

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
add missing header

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
generalize RetainableByteBuffer

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
enable thread-local cache on the pool

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
do not handle null retainableByteBufferPool

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
make http client register RetainableByteBufferPool

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
fix concurrency

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
Merge remote-tracking branch 'origin/jetty-10.0.x' into jetty-10.0.x-6322-UseRetainableByteBuffer

lorban committed:
add & fix javadoc

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
improve javadoc

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
pre-allocate buckets

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
add helper method that either finds or wraps a memory pool

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
handle review comments

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
handle review comments

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
rename

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
rename

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
change bean lookup

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
fix checkstyle

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
fix error message

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
add memory calculation + clear + tests

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
add available memory calculation + tests

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
rename

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
add buffer eviction mechanism

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
use atomics to keep track of memory

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
simplify buffer eviction

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
do not expose retain counter in API anymore

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
set the initial retain counter to 0 + improve javadoc

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
code alignment

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
move findOrAdapt to interface

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
remove magic constant

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
move findOrAdapt to interface

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
add retain test

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
make getRequestBuffer private

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
simplify if/else

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
set the initial retain counter to 0 + improve javadoc

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
force releasing the buffer on exception

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
fix checkstyle

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
add back retain/release safety check

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
fix tests visibility

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
review comments

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
review comments

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
rename to ArrayRetainableByteBufferPool

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
do bound checking in an atomic function to avoid exposing illegal values

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
fix checkstyle

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

gregw committed:
Improve #6322 log buckets in RetainableByteBufferPool

Add a Log2 bucket size imple for ArrayRetainableByteBufferPool

Signed-off-by: Greg Wilkins <gregw@webtide.com>

gregw committed:
Improve #6322 log buckets in RetainableByteBufferPool

Add a Log2 bucket size impl for ArrayRetainableByteBufferPool

Signed-off-by: Greg Wilkins <gregw@webtide.com>

gregw committed:
Cleanups

Use Bucket everywhere
Use -1 to indicate defaults

Signed-off-by: Greg Wilkins <gregw@webtide.com>

lorban commented on jetty-io/src/main/java/org/eclipse/jetty/io/ArrayRetainableByteBufferPool.java:
```diff
@@ -311,4 +334,45 @@ private void evict(boolean direct, long excess)
+                maxBucketSize,
+                maxHeapMemory,
+                maxDirectMemory,
+                TypeUtil::log2NextPowerOf2,
```
I would not change the existing buckets-sizing algorithm without measurements. Heck, I'm not even convinced making it configurable (even only in code via an internal API) is a good idea.

We currently have little visibility over how buckets are sized in the wild so I'm worried that changing this may lead to adverse effects.

gregw replied:
We need to make changes like this PR so we can see how buckets are sized in the wild.     I'm definitely not proposing that `LogBuckets` be used until we see the data, but making the pool Dumpable, sampleable and extensible are all worthwhile things.

sbordet replied:
I'm not against the power of 2 idea, but perhaps it should be stripped out to maintain, as you say, dumpability and statistics.

gregw commented:
@lorban The intention is not to change the existing behaviour.  I don't think I have and the CI failure appears to be exactly that.... a CI failure.  Will redo the CI tests to see.

But I'm a strong advocate of making this class extensible, even if we don't add  the badly named `LogBuckets`.  Unless the class is extensible then we won't ever be able to do any comparisons.
Also I'll include in this making the pool dumpable so that we can see a lot more detail of how the buckets are being used (regardless of allocation).  We can then get some dumps from real live servers and do a calculation to see how alternate bucket strategies might affect the pattern we see.

gregw committed:
Improve #6322 log buckets in RetainableByteBufferPool

Dumpable

Signed-off-by: Greg Wilkins <gregw@webtide.com>

gregw committed:
Merge branch 'jetty-10.0.x-6322-UseRetainableByteBuffer-LogBuckets' of github.com:eclipse/jetty.project into jetty-10.0.x-6322-UseRetainableByteBuffer-LogBuckets

Signed-off-by: Greg Wilkins <gregw@webtide.com>

gregw committed:
Improve #6322 log buckets in RetainableByteBufferPool

Sampled Pool

Signed-off-by: Greg Wilkins <gregw@webtide.com>

gregw committed:
Merge remote-tracking branch 'origin/jetty-10.0.x-6322-UseRetainableByteBuffer' into jetty-10.0.x-6322-UseRetainableByteBuffer-LogBuckets

gregw commented:
@sbordet @lorban ignoring the `LogBucket` class itself, I think this PR has some good improvements for:
- code clarity with the `Bucket` sub class 
- setting defaults (not duplicated in constructors)
- dumpability so we can see the distribution in buckets and how buffers are being used
- extensibility, we can have a `LogBuckets` if it proves necessary.

gregw commented:
@sbordet The problem with stripping it out, is that we then have no way of actually testing that the class is really extensible.    There were a lot of little assumptions in the code that depended on linear buckets. This PR fixes them, but if we took out  LogBuckets, it would be easy for a change to revert to having such a assumptions.    I could move LogBuckets to test???  But I really don't see any harm having it available, as it will allow us to test with it in real world deployments.

sbordet commented on jetty-util/src/main/java/org/eclipse/jetty/util/TypeUtil.java:
```diff
@@ -800,4 +800,42 @@ public T next()
+     * @param v An integer &gt; 0 and &lt;= half of {@link Integer#MAX_VALUE}
+     * @return The next power of two that is equal too or larger than the passed integer.
+     */
+    public static int nextPowerOf2(int v)
```
Can you please check whether you can do this with `Long.highestOneBit()`?

sbordet commented on jetty-io/src/main/java/org/eclipse/jetty/io/ArrayRetainableByteBufferPool.java:
```diff
@@ -311,4 +359,51 @@ private void evict(boolean direct, long excess)
+     * uses buckets of buffers that increase in size by a power of
+     * 2 (eg 1k, 2k, 4k, 8k, etc.).
+     */
+    public static class LogBuckets extends ArrayRetainableByteBufferPool
```
I like better `PowerOfTwoStrategy` or `ExponentialStrategy`.

gregw committed:
Merge remote-tracking branch 'origin/jetty-10.0.x-6322-UseRetainableByteBuffer' into jetty-10.0.x-6322-UseRetainableByteBuffer-LogBuckets

Updates from review

Signed-off-by: Greg Wilkins <gregw@webtide.com>

lorban commented on jetty-io/src/main/java/org/eclipse/jetty/io/RetainableByteBufferPool.java:
```diff
@@ -55,4 +60,56 @@ static RetainableByteBufferPool findOrAdapt(Container container, ByteBufferPool
+    {
+        private final RetainableByteBufferPool _pool;
+        private final SampleStatistic _heap = new SampleStatistic();
+        private final SampleStatistic _direct = new SampleStatistic();
```
I think those two counters are going to be a bit meaningless at the pool layer. IMHO it would be a lot more informative to move them to the `Bucket` layer, maybe by creating a `DumpableBucket` class.

gregw replied:
Hmmm good test for extensibility to have SampledBuckets.    Will try....

gregw replied:
Actaully,   thinking about it, we can just run with a small factor and then we will effectively get a histogram of the acquired sizes.    So maybe this class is not needed after all.

lorban reviewed (changes requested):
I quite like the idea behind this change. More visibility and extensibility is always good to have. I just think that the buckets should be sampled instead of the pool.

sbordet commented on jetty-io/src/main/java/org/eclipse/jetty/io/ArrayRetainableByteBufferPool.java:
```diff
@@ -52,48 +57,74 @@ public ArrayRetainableByteBufferPool(int minCapacity, int factor, int maxCapacit
+            bucketCapacity = i -> (i + 1) * f;
 
+        int length = bucketIndexFor.apply(maxCapacity) + 1;
         @SuppressWarnings("unchecked")
```
Unnecessary suppression.

sbordet commented on jetty-io/src/main/java/org/eclipse/jetty/io/ArrayRetainableByteBufferPool.java:
```diff
@@ -52,48 +57,74 @@ public ArrayRetainableByteBufferPool(int minCapacity, int factor, int maxCapacit
         @SuppressWarnings("unchecked")
-        Pool<RetainableByteBuffer>[] directArray = new Pool[length];
+        Bucket[] directArray = new Bucket[length];
         @SuppressWarnings("unchecked")
```
Unnecessary suppression.

sbordet commented on jetty-io/src/main/java/org/eclipse/jetty/io/ArrayRetainableByteBufferPool.java:
```diff
@@ -52,48 +57,74 @@ public ArrayRetainableByteBufferPool(int minCapacity, int factor, int maxCapacit
             maxCapacity = 64 * 1024;
-        if ((maxCapacity % _factor) != 0 || _factor >= maxCapacity)
+
+        final int f = factor <= 0 ? 1024 : factor;
```
Drop `final`, not necessary.

sbordet commented on jetty-io/src/main/java/org/eclipse/jetty/io/ArrayRetainableByteBufferPool.java:
```diff
@@ -311,4 +358,51 @@ private void evict(boolean direct, long excess)
+     * uses buckets of buffers that increase in size by a power of
+     * 2 (eg 1k, 2k, 4k, 8k, etc.).
+     */
+    public static class ExponentialPool extends ArrayRetainableByteBufferPool
```
Not sure we want to make this `public`, as it binds us to maintain it, etc.

gregw replied:
I want to make it public so we can easily try using it in real deployments.    
But as there appears to be resistance to this idea, I'll move to test and we can used the sampled one to gather data to see if `ExponentialPool` is worthwhile or not.

sbordet commented on jetty-io/src/main/java/org/eclipse/jetty/io/RetainableByteBufferPool.java:
```diff
@@ -55,4 +60,56 @@ static RetainableByteBufferPool findOrAdapt(Container container, ByteBufferPool
+     * @see SampleStatistic
+     */
+    @ManagedObject
+    class SampledPool implements RetainableByteBufferPool, Dumpable
```
This class is implicitly `public`, which I don't think it's a good idea.

gregw replied:
We want at least this one to be public, so we can run real deployments and capture data about the pool usage.   You can't simultaneously argue that `ExponentialPool` is not justified by any data and then prevent anything that allows the data to be collected!

sbordet commented on jetty-io/src/main/java/org/eclipse/jetty/io/RetainableByteBufferPool.java:
```diff
@@ -55,4 +60,56 @@ static RetainableByteBufferPool findOrAdapt(Container container, ByteBufferPool
+        }
+
+        @ManagedAttribute("Heap acquire samples")
+        public SampleStatistic getHeapSample()
```
Unfortunately `SampleStatistic` is not manageable, so this JMX exposure results in no data being exposed.
I think we need either more specific getters, or MXBean-ify `SampleStatistic` which may not be easy.

I would make this class package private in `ArrayRBBP` for now, and eventually open it up later.

gregw replied:
If it is private, then we can't use it to gather any info.
Anyway, I'm going to look at sampling on buckets rather than on acquire.

sbordet commented on jetty-io/src/test/java/org/eclipse/jetty/io/ArrayRetainableByteBufferPoolTest.java:
```diff
@@ -313,4 +315,33 @@ public void testAcquireRelease()
+            capacity = capacity * 2;
+        }
+
+        System.err.println(pool.dump());
```
Remove this?

gregw committed:
Merge remote-tracking branch 'origin/jetty-10.0.x-6322-UseRetainableByteBuffer' into jetty-10.0.x-6322-UseRetainableByteBuffer-LogBuckets

Updates from review

Signed-off-by: Greg Wilkins <gregw@webtide.com>

gregw commented:
@lorban @sbordet I have removed `SampledPool` as we get a reasonable histogram of acquire size by using the normal linear pool and dumping the bucket sizes.
I have added a bit more info into the bucket dumps, but as the sampling is in a race, it may be of marginal value on a server underload, but will be OK if dumped after a load test.

I have moved the ExponentialPool to be an inner class of the test.   So process from here is to release this.... ask some real clients do give us some server dumps; analyse the "histograms"; if an ExponentialPool looks like it will work, then re-release with that class public and test.

gregw committed:
Merge remote-tracking branch 'origin/jetty-10.0.x-6322-UseRetainableByteBuffer' into jetty-10.0.x-6322-UseRetainableByteBuffer-LogBuckets

Updates from review

Signed-off-by: Greg Wilkins <gregw@webtide.com>

gregw commented:
@lorban @sbordet nudge

lorban commented on jetty-io/src/main/java/org/eclipse/jetty/io/ArrayRetainableByteBufferPool.java:
```diff
@@ -311,4 +356,53 @@ private void evict(boolean direct, long excess)
+                }
+            }
+
+            return String.format("%s{capacity=%d,inuse=%d(%d%%),used=%d%%}",
```
I'm not sure the `used` calculation will report anything meaningful, a more useful metric would be to calculate the mean 'waste' (i.e.: the difference between the requested size and the returned buffer's capacity) but that would require some extra processing.

I think we should leave this as-is and eventually revisit that later on.

gregw replied:
Agreed. We are kind of thwarted by the fact that we have no idea if a buffer is in fill or flush mode.
At least with this metric, it is sampling buffers that have a pointer between 0 and capacity, which is at least a hint of partial usage.

lorban reviewed (approved):
Besides the checkstyle failure, LGTM.

gregw committed:
Merge remote-tracking branch 'origin/jetty-10.0.x-6322-UseRetainableByteBuffer' into jetty-10.0.x-6322-UseRetainableByteBuffer-LogBuckets

fix checkstyle

Signed-off-by: Greg Wilkins <gregw@webtide.com>

gregw committed:
Merge remote-tracking branch 'origin/jetty-10.0.x-6322-UseRetainableByteBuffer' into jetty-10.0.x-6322-UseRetainableByteBuffer-LogBuckets

fix checkstyle
removed useless metric

Signed-off-by: Greg Wilkins <gregw@webtide.com>

gregw commented:
@sbordet nudge!!!

gregw commented:
Closed in favour of #6538
